cpp/src/io/utilities/multibuffer_memset.cu

```diff
@@ -0,0 +1,145 @@
+  rmm::device_uvector<std::size_t> offsets(bufs.size() + 1, _stream, _mr); 
+
+  // copy bufs into gpu and then get sizes from there (cudf detail function make device vector async)
+  auto gpu_bufs = cudf::detail::make_device_uvector_async(bufs, _stream, _mr);
+
+  // get a vector with the sizes of all buffers
+  auto buf_count_iter = cudf::detail::make_counting_transform_iterator(
+    0, 
+    cuda::proclaim_return_type<std::size_t>(
+      [gpu_bufs = gpu_bufs.data(), bytes_per_task, num_bufs] __device__(cudf::size_type i) {
+        size_t temp = cudf::util::round_up_safe(gpu_bufs[i].size(), bytes_per_task);
+        return i >= num_bufs ? 0 : temp / bytes_per_task;
+      }
+    )
+  );
+
+  // fill up offsets buffer using exclusive scan
+  thrust::exclusive_scan(rmm::exec_policy(_stream, _mr), buf_count_iter, buf_count_iter + bufs.size() + 1, offsets.begin(), 0);
+
+  // the total number of tasks is the last number in the offsets vector
+  size_t const total_tasks = offsets.back_element(_stream);
+
+  // declaring list of tasks to pass onto cuda kernel
+  rmm::device_uvector<memset_task> tasks(total_tasks, _stream, _mr); 
+
+  // fill up task lists based on buffer values
+  thrust::transform(
+    rmm::exec_policy(_stream, _mr),
+    thrust::make_counting_iterator<std::size_t>(0), 
+    thrust::make_counting_iterator<std::size_t>(total_tasks), 
+    tasks.begin(), 
+    cuda::proclaim_return_type<memset_task>(
+      [offsets = offsets.data(), offset_size = num_bufs + 1, gpu_bufs = gpu_bufs.data(), bytes_per_task] __device__(cudf::size_type task) {
+        auto buf_idx = thrust::upper_bound(thrust::seq, offsets, offsets + offset_size, task);
+        if (*buf_idx > task) {buf_idx -= 1;}
+        size_t start = (task - *buf_idx) * bytes_per_task;
+        size_t end = (start + bytes_per_task) <= gpu_bufs[buf_idx - offsets].size() ? start + bytes_per_task : gpu_bufs[buf_idx - offsets].size();
+        memset_task ret;
+        ret.size = (end - start) / 8;
+        ret.data = (uint64_t *)(gpu_bufs[buf_idx - offsets].data() + start);
+        return ret;
+      }
+    )
+  );
+
+  // launch cuda kernel
+  if (total_tasks != 0) {
+    memset_kernel<<<total_tasks, threads_per_block>>>(tasks.data(), value);
+  }
+}
+
+} // namespace detail
+
+void multibuffer_memset(std::vector<cudf::device_span<uint8_t>> & bufs, 
+                        int8_t const value,
+                        rmm::cuda_stream_view _stream,
+                        rmm::device_async_resource_ref _mr
+                        )
+{
+  detail::multibuffer_memset(bufs, value, _stream, _mr, 1024 * 128, 256);
+}
+
+void multibuffer_memset_validity(std::vector<cudf::device_span<uint8_t>> & bufs, 
+                        int8_t const value,
+                        rmm::cuda_stream_view _stream,
+                        rmm::device_async_resource_ref _mr
+                        )
+{
+  detail::multibuffer_memset(bufs, value, _stream, _mr, 1024, 256);
```

Comment:
1024 bytes, 256 threads is a strange combination. We're writing 8 bytes per thread at a time, so we're never actually using more than 128 threads in the block. Do you have some numbers here showing a plot of the results of various params?  I think it would be useful to see a compute-sanitizer result from here.  Unless I'm mistaken we should see something like at most 50% actual warp utilization.